contracts/modules/cover/Cover.sol

```diff
@@ -186,343 +185,186 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon, ReentrancyGuard {
-          allocation.coverAmountInNXM,
-          lastCoverSegment.period,
-          uint(lastCoverSegment.gracePeriodInDays) * 1 days,
-          // TODO globalCapacityRatio and capacityReductionRatio need to be stored at cover buy
+          coveredAmountInNXM,
+          params.period
+        );
+
+        AllocationRequestConfig memory config = AllocationRequestConfig(
+          uint(_productTypes[product.productType].gracePeriodInDays) * 1 days,
           globalCapacityRatio,
           product.capacityReductionRatio,
-          lastCoverSegment.globalRewardsRatio
+          globalRewardsRatio,
+          GLOBAL_MIN_PRICE_RATIO
         );
 
-        stakingPool(allocation.poolId).deallocateStake(
-          request,
-          lastCoverSegment.start,
-          allocation.premiumInNXM
+        premiumInNXM += stakingPool(allocationRequests[i].poolId).allocateCapacity(allocationRequest, config);
+        totalCoverAmountInNXM += coveredAmountInNXM;
+
+        coverSegmentAllocations[coverId][segmentId].push(
+          PoolAllocation(
+            allocationRequests[i].poolId,
+            SafeUintCast.toUint96(coveredAmountInNXM)
+          )
         );
+      }
 
-        coverSegmentAllocations[coverId][lastCoverSegmentIndex][i].premiumInNXM =
-          allocation.premiumInNXM
-          * (lastCoverSegment.period - remainingPeriod)
-          / lastCoverSegment.period;
+      uint coverAmountInCoverAsset = totalCoverAmountInNXM * nxmPriceInCoverAsset / ONE_NXM;
 
-        // Compute NXM rewards to be rolled back
-        uint deallocatedRewardsInNXM = allocation.premiumInNXM
-          * remainingPeriod / lastCoverSegment.period
-          * lastCoverSegment.globalRewardsRatio / REWARD_DENOMINATOR;
+      _coverSegments[coverId].push(
+        CoverSegment(
+          coverAmountInCoverAsset.toUint96(), // amount
+          uint32(block.timestamp + 1), // start
+          SafeUintCast.toUint32(params.period), // period
+          _productTypes[_products[params.productId].productType].gracePeriodInDays,
+          globalRewardsRatio
+        )
+      );
 
-        tokenController().burnStakingPoolNXMRewards(deallocatedRewardsInNXM, allocation.poolId);
-      }
-    }
+      retrievePayment(
+        premiumInNXM,
+        params.paymentAsset,
+        params.maxPremiumInAsset,
+        params.commissionRatio,
+        params.commissionDestination
+      );
 
-    refundInCoverAsset = lastCoverSegment.priceRatio
-      * lastCoverSegment.amount
-      / PRICE_DENOMINATOR
-      * remainingPeriod
-      / MAX_COVER_PERIOD;
+      // TODO: implement using buckets
+      totalActiveCoverInAsset[params.coverAsset] += coverAmountInCoverAsset;
 
-    // Edit the last cover segment so it ends at the current block by writing to storage directly
-    _coverSegments[coverId][lastCoverSegmentIndex].period = lastCoverSegment.period - remainingPeriod;
+      emit CoverEdited(coverId, params.productId, segmentId, msg.sender, params.ipfsData);
```

Comment:
Is it worth it to conditionally emit ipfsData here, like we do in `setProducts()`?

Reply:
Since each cover segment semantically makes sense and a new one is created on each edit - I would explicitly emit it every time in this case.